File: packages/sdk-java/src/test/java/com/cdpus/sdk/CdpClientTest.java
package com.cdpus.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

final class CdpClientTest {
  @Test
  void postsTrackPayloadToTrackEndpoint() {
    FakeTransport transport = new FakeTransport(202);
    CdpClient client = CdpClient.builder().writeKey("wk_us").endpoint("https://api.example.com").transport(transport).build();

    client.track("anon_1", "Signed Up", Map.of("plan", "growth"));
    client.flush();

    assertEquals("https://api.example.com/v1/track", transport.urls.get(0));
    assertTrue(transport.bodies.get(0).contains("\"writeKey\":\"wk_us\""));
    assertTrue(transport.bodies.get(0).contains("\"events\""));
  }

  @Test
  void flushesAtThresholdAndCloseFlushes() {
    FakeTransport transport = new FakeTransport(202, 202);
    CdpClient client = CdpClient.builder().writeKey("wk").flushAt(2).transport(transport).build();
    client.track("a", "One", Map.of());
    client.identify("a", Map.of("email", "buyer@example.com"), "u1");
    assertEquals(1, transport.bodies.size());
    client.track("a", "Buffered", Map.of());
    client.close();
    assertEquals(2, transport.bodies.size());
  }

  @Test
  void retriesFiveHundredButNotFourHundred() {
    FakeTransport retry = new FakeTransport(500, 202);
    CdpClient.builder().writeKey("wk").flushAt(1).transport(retry).retryDelayMs(0).build().track("a", "Retry", Map.of());
    assertEquals(2, retry.bodies.size());

    FakeTransport bad = new FakeTransport(400, 202);
    CdpClient client = CdpClient.builder().writeKey("wk").flushAt(1).transport(bad).build();
    assertThrows(CdpException.class, () -> client.track("a", "Bad", Map.of()));
    assertEquals(1, bad.bodies.size());
  }

  static final class FakeTransport implements Transport {
    final List<Integer> statuses;
    final List<String> urls = new ArrayList<>();
    final List<String> bodies = new ArrayList<>();

    FakeTransport(Integer... statuses) {
      this.statuses = new ArrayList<>(List.of(statuses));
    }

    @Override
    public int send(String url, String jsonBody) throws IOException {
      urls.add(url);
      bodies.add(jsonBody);
      return statuses.isEmpty() ? 202 : statuses.remove(0);
    }
  }
}

File: packages/sdk-php/tests/CdpClientTest.php
<?php

declare(strict_types=1);

namespace CdpUs\Tests;

use CdpUs\CdpClient;
use CdpUs\TransportException;
use PHPUnit\Framework\TestCase;

final class CdpClientTest extends TestCase
{
    public function testPostsTrackPayloadToTrackEndpoint(): void
    {
        $transport = new FakeTransport();
        $client = new CdpClient('wk_us', 'https://api.example.com', 20, $transport);

        $client->track('anon_1', 'Signed Up', ['plan' => 'growth']);
        $client->flush();

        self::assertSame('https://api.example.com/v1/track', $transport->requests[0]['url']);
        self::assertSame('wk_us', $transport->lastPayload()['writeKey']);
        self::assertSame('track', $transport->lastPayload()['events'][0]['type']);
    }

    public function testFlushesWhenBufferReachesFlushAt(): void
    {
        $transport = new FakeTransport();
        $client = new CdpClient('wk_us', 'http://localhost:8110', 2, $transport);

        $client->track('anon_1', 'One');
        $client->identify('anon_1', ['email' => 'buyer@example.com'], 'user_1');

        self::assertCount(1, $transport->requests);
        self::assertCount(2, $transport->lastPayload()['events']);
    }

    public function testRetriesFiveHundredThenSucceeds(): void
    {
        $transport = new FakeTransport([500, 202]);
        $client = new CdpClient('wk_us', 'http://localhost:8110', 1, $transport, retryDelayMs: 0);

        $client->track('anon_1', 'Retry');

        self::assertCount(2, $transport->requests);
    }

    public function testDoesNotRetryFourHundred(): void
    {
        $transport = new FakeTransport([400, 202]);
        $client = new CdpClient('wk_us', 'http://localhost:8110', 1, $transport);

        $this->expectException(TransportException::class);
        $this->expectExceptionMessage('CDP rejected 400');

        try {
            $client->track('anon_1', 'Bad');
        } finally {
            self::assertCount(1, $transport->requests);
        }
    }

    public function testCloseFlushesBufferedEvents(): void
    {
        $transport = new FakeTransport();
        $client = new CdpClient('wk_us', 'http://localhost:8110', 20, $transport);

        $client->track('anon_1', 'Buffered');
        $client->close();

        self::assertCount(1, $transport->requests);
        self::assertSame('Buffered', $transport->lastPayload()['events'][0]['event']);
    }
}
